### PR TITLE
configs: Update configs in configs/example/gem5_library for v25.0

### PR DESCRIPTION
**Pull request by erin-le**

### CODE REVIEW

File: configs/example/gem5_library/arm-demo-ubuntu-run.py [hunk 1]
```diff
@@ -45,7 +45,8 @@
 from gem5.simulate.simulator import Simulator
 from gem5.utils.requires import requires
 
-# This runs a check to ensure the gem5 binary interpreting this file is compiled to include the ARM ISA.
+# This runs a check to ensure the gem5 binary interpreting this file is
+# compiled to include the ARM ISA.
```
Comment (powerjg) [inline configs/example/gem5_library/arm-demo-ubuntu-run.py]: I don't actually think we need these checks in the example files. I believe that it will have duplicate checks when importing the `ArmBoard`. Let me know if I'm wrong.

File: configs/example/gem5_library/arm-ubuntu-run.py [hunk 2]
```diff
@@ -105,27 +108,45 @@
...
-        # exit event.
-        ExitEvent.EXIT: exit_event_handler()
-    },
-)
+
+# You can inherit from either the class that handles a certain hypercall by
+# default, or inherit directly from ExitHandler and specify a hypercall number.
+# See src/python/gem5/simulate/exit_handler.py for more information on which
+# behaviors map to which hypercalls, and what the default behaviors are.
+class CustomKernelBootedExitHandler(KernelBootedExitHandler):
```
Comment (powerjg) [inline configs/example/gem5_library/arm-ubuntu-run.py:113]: A *HUGE* thank you for doing this!!

File: configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py [hunk 3]
```diff
@@ -87,6 +87,14 @@
 # We get the pre-taken checkpoint from gem5-resources. This checkpoint
 # was taken from running this gem5 configuration script,
 # configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py
+
+# gem5 v25.0: `riscv-hello` on gem5 resources needs to be updated to say it
+# is compatible with v25.0
+# The description of the riscv-hello-example-checkpoint says it is only
+# compatible with gem5 v22.0, but the raw JSON shows that it is compatible
+# with gem5 v24.1 and v25.0
```
Comment (powerjg) [inline configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py]: Is this a "TODO"? If so, make an issue.

File: configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py [hunk 4]
```diff
@@ -92,30 +96,36 @@
...
+    @overrides(ScheduledExitEventHandler)
+    def _exit_simulation(self) -> bool:
+        return True
+
 
 # Lastly we run the simulation.
 max_ticks = 10**6
-simulator = Simulator(board=board, full_system=False, max_ticks=max_ticks)
-simulator.run()
+m5.scheduleTickExitAbsolute(max_ticks)
```
Comment (powerjg) [inline configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py]: Two things:
1. Why are we using `m5.scheduleTickAbsolute`? Doesn't the `Simulator` have a function for this?
2. Use `1_000_000` instead of a floating point number.

File: configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py [hunk 5]
```diff
@@ -110,33 +110,43 @@
...
+#         simpoint_interval=1000000,
+#         simpoint_list=[2, 3, 4, 15],
+#         weight_list=[0.1, 0.2, 0.4, 0.3],
+#         warmup_interval=1000000,
+#     ),
+# )
+
+# The workload x86-print-this-15000-with-simpoints, version 2.0.0 is available
+# for gem5 versions 23.1 through 24.1. We may need to create another version
+# of the workload that throws a hypercall (see comment below)
```
Comment (powerjg) [inline configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py]: Make an issue, please. If you want to go ahead and merge this PR, reference the issue here.

File: configs/example/gem5_library/checkpoints/simpoints-se-restore.py [hunk 6]
```diff
@@ -106,58 +102,76 @@
...
 
-# Schedule a MAX_INSTS exit event before the simulation begins the
-# schedule_max_insts function only schedule event when the instruction length
-# is greater than 0.
-# In here, it schedules an exit event for the first SimPoint's warmup
-# instructions
-simulator.schedule_max_insts(board.get_simpoint().get_warmup_list()[0])
+# Here, m5.scheduleTickExitAbsolute schedules an exit event for the first
+# SimPoint's warmup instructions
+m5.scheduleTickExitAbsolute(board.get_simpoint().get_warmup_list()[0])
```
Comment (powerjg) [inline configs/example/gem5_library/checkpoints/simpoints-se-restore.py]: Again, why not use `schedule_max_insts()`?

File: configs/example/gem5_library/exit_handling/user-exit-handler.py [hunk 7]
```diff
@@ -72,17 +71,19 @@
 )
 
 # Set the workload
+# The binary on gem5 resources needs to be updated to say that it is
+# compatible with gem5 v25.0
```
Comment (powerjg) [inline configs/example/gem5_library/exit_handling/user-exit-handler.py]: To do?

File: configs/example/gem5_library/looppoints/create-looppoint-checkpoints.py [hunk 8]
```diff
@@ -117,6 +117,10 @@
 dir = Path(args.checkpoint_path)
 dir.mkdir(exist_ok=True)
 
+# The workload may have to be updated so it throws a hypercall instead of
+# hypercall 0. Then, this config will have to be modified so it has a handler
+# for that hypercall
```
Comment (powerjg) [inline configs/example/gem5_library/looppoints/create-looppoint-checkpoints.py]: Issue?

File: configs/example/gem5_library/looppoints/restore-looppoint-checkpoint.py [hunk 9]
```diff
@@ -115,6 +116,7 @@
     cache_hierarchy=cache_hierarchy,
 )
 
+# This workload may have to be updated to throw hypercalls
```
Comment (powerjg) [inline configs/example/gem5_library/looppoints/restore-looppoint-checkpoint.py]: To do? Issue?

File: configs/example/gem5_library/memory_traffic.py [hunk 10]
```diff
@@ -99,6 +99,9 @@ def generator_factory(
 # We use the Test Board. This is a special board to run traffic generation
 # tasks. Can replace the cache_hierarchy with any hierarchy to simulate the
 # cache as well as the memory
+
+# Need to test this; will this work with cache_hierarchy as None instead of
+# NoCache?
```
Comment (powerjg) [inline configs/example/gem5_library/memory_traffic.py]: The test board should know how to deal with `None` cache instead of `NoCache`.  `NoCache` is actually a cache hierarchy, but that doesn't have any caches. `None` actually means that the generators are directly connected to the memory.

File: configs/example/gem5_library/riscv-fs.py [hunk 11]
```diff
@@ -51,6 +51,10 @@
 from gem5.simulate.simulator import Simulator
 from gem5.utils.requires import requires
 
+# Will we continue to support the `riscv-disk-img` resource? If not, should this
+# config script be deprecated, given that it would be very similar to
+# `riscv-ubuntu-run.py` if we updated the workload to a more recent one?
```
Comment (powerjg) [inline configs/example/gem5_library/riscv-fs.py]: Yes. I think this should be deprecated. That should be a separate PR, though.

File: configs/example/gem5_library/riscv-rvv-example.py [hunk 12]
```diff
@@ -38,14 +39,15 @@
...
 
 # Run the simulation
-./build/RISCV/gem5.opt configs/example/gem5_library/riscv-rvv-example.py \
+./build/ALL/gem5.opt configs/example/gem5_library/riscv-rvv-example.py \
     [-c CORES] [-v VLEN] [-e ELEN] <resource>
 
 """
-
+# Should I add the UC Davis copyright if I modify this? Or do changes have to
+# be above a certain size/amount to add the copyright?
```
Comment (powerjg) [inline configs/example/gem5_library/riscv-rvv-example.py]: Use your best judgement. In this case, I don't think it matters.

File: configs/example/gem5_library/riscvmatched-fs.py [hunk 13]
```diff
@@ -59,28 +65,36 @@
...
+
+# instantiate the RISCV Matched board with default parameters
 board = RISCVMatchedBoard(
     clk_freq="1.2GHz",
     l2_size="2MiB",
     is_fs=True,
 )
+# We previously tried to update this config script to boot Ubuntu 24.04, but
+# reverted back to Ubuntu 20.04 because we ran into issues. This will have
+# to be tested again.
```
Comment (powerjg) [inline configs/example/gem5_library/riscvmatched-fs.py]: To do or create an issue.

File: configs/example/gem5_library/x86-gapbs-benchmarks.py [hunk 14]
```diff
@@ -64,19 +61,28 @@
...
     kvm_required=True,
 )
 
 parser = argparse.ArgumentParser(
-    description="An example configuration script to run the gapbs benchmarks."
+    description="An example configuration script to run the GAPBS benchmarks."
 )
 
+# This suite on the gem5 Resources website should be updated to use the latest
+# GAPBS benchmarks. For now, I will modify this script not to use the suite
```
Comment (powerjg) [inline configs/example/gem5_library/x86-gapbs-benchmarks.py]: To do or issue?

File: configs/example/gem5_library/memory_traffic.py [hunk 15]
```diff
@@ -99,6 +99,9 @@ def generator_factory(
...
 board = TestBoard(
     clk_freq="1GHz",  # Ignored for these generators
     generator=generator,  # We pass the traffic generator as the processor.
```
Comment (erin-le) [inline configs/example/gem5_library/memory_traffic.py]: Need to test this; will it work with `cache_hierarchy=None` instead of `cache_hierarchy=NoCache()`?

File: configs/example/gem5_library/x86-gapbs-benchmarks.py [hunk 16]
```diff
@@ -139,40 +145,72 @@
 )
 
 # Here we set the FS workload, i.e., gapbs benchmark program
-# After simulation has ended you may inspect
-# `m5out/system.pc.com_1.device` to the stdout, if any.
+# After simulation has ended you may inspect `m5out/system.pc.com_1.device`
```
Comment (powerjg) [inline configs/example/gem5_library/x86-gapbs-benchmarks.py]: I think this path is wrong. I think it should be `m5out/board.pc.com_1.device`

File: configs/example/gem5_library/x86-gapbs-benchmarks.py [hunk 17]
```diff
@@ -139,40 +145,72 @@
...
+#     global start_tick
+#     start_tick = m5.curTick()
+#     processor.switch()
+#     yield False  # E.g., continue the simulation.
+
+
+# def handle_workend():
+#     print("Dump stats at the end of the ROI!")
+#     m5.stats.dump()
+#     yield True  # Stop the simulation. We're done.
```
Comment (powerjg) [inline configs/example/gem5_library/x86-gapbs-benchmarks.py]: delete?

File: configs/example/gem5_library/x86-gapbs-benchmarks.py [hunk 18]
```diff
@@ -139,40 +145,72 @@
...
 
-simulator = Simulator(
-    board=board,
-    on_exit_event={
-        ExitEvent.WORKBEGIN: handle_workbegin(),
-        ExitEvent.WORKEND: handle_workend(),
-    },
-)
+        m5.stats.dump()
+        # This is a hacky way to keep the script's ending print statement.
```
Comment (powerjg) [inline configs/example/gem5_library/x86-gapbs-benchmarks.py]: Don't worry about this. It can be dropped.

File: configs/example/gem5_library/x86-global-inst-tracker.py [hunk 19]
```diff
@@ -113,6 +113,7 @@
     memory=memory,
     cache_hierarchy=cache_hierarchy,
 )
+# The workload needs to be updated to use hypercalls
```
Comment (powerjg) [inline configs/example/gem5_library/x86-global-inst-tracker.py]: Issue?

File: configs/example/gem5_library/x86-npb-benchmarks.py [hunk 20]
```diff
@@ -99,14 +106,15 @@
...
-    type=int,
-    help="Optionally put the maximum number of ticks to execute during the "
-    "ROI. It accepts an integer value.",
-)
+# parser.add_argument(
+#     "--ticks",
+#     type=int,
+#     help="Optionally put the maximum number of ticks to execute during the "
+#     "ROI. It accepts an integer value."
+# )
```
Comment (powerjg) [inline configs/example/gem5_library/x86-npb-benchmarks.py]: I think you can delete all of the commented code in this file.

File: configs/example/gem5_library/x86-parsec-benchmarks.py [hunk 21]
```diff
@@ -63,6 +63,11 @@
 from gem5.simulate.simulator import Simulator
 from gem5.utils.requires import requires
 
+# It doesn't look like we have any up to date Parsec resources. The
```
Comment (powerjg) [inline configs/example/gem5_library/x86-parsec-benchmarks.py]: Deprecate this script.

**Review by powerjg — COMMENTED** [review]
First.. **amazing job!!** You have done a great job of cleaning this up. Way to go 🥇 

Have all (any) of these been tested? We should figure out a way to communicate what has been tested.

Also, everything that is something to do in the future should either be marked as "to do" before this is merged or an issue should be created (and there should be cross links from the issue to the code and vice versa).

**Comment by erin-le** [comment]
@powerjg: Hey Jason, thanks for the review. I'm still in the progress of cleaning up these scripts, so there will be more changes shortly.

None of these scripts have been tested yet. I'll also extract the comments about things that need to be updated into an issue, as you mentioned above.

File: configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py (hunk 4 above)
Comment (erin-le) [reply configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py]: Due to the [comment in the hypercall 6 handler](https://github.com/gem5/gem5/blob/654fa090c17c82eeaefff7435929528e858afcf6/src/python/gem5/simulate/exit_handler.py#L184), I was always under the impression that only `m5.scheduleTickExitFromCurrent` or `m5.scheduleTickExitAbsolute` could be used to trigger a hypercall 6 exit. Can `Simulator.schedule_max_insts()` also trigger hypercall 6? From the comment in `schedule_max_insts()`, it looks like it triggers an ExitEvent.MAX_INSTS exit event.

File: configs/example/gem5_library/arm-demo-ubuntu-run.py (hunk 1 above)
Comment (erin-le) [reply configs/example/gem5_library/arm-demo-ubuntu-run.py]: It looks like the demo boards and `ArmBoard` have the `requires(...)` check, but `RiscvBoard` and `X86Board` use `if processor.get_isa() != ISA.[ISA]` instead. I'll remove the `requires(...)` check from the configs that use the demo boards and `ArmBoard`.

Should I also remove the `requries` check for configs that use `RiscvBoard` and `X86Board`?

File: configs/example/gem5_library/memory_traffic.py (hunk 15 above)
Comment (powerjg) [reply configs/example/gem5_library/memory_traffic.py]: It should work with both.

File: configs/example/gem5_library/arm-demo-ubuntu-run.py (hunk 1 above)
Comment (powerjg) [reply configs/example/gem5_library/arm-demo-ubuntu-run.py]: It would be good for it to be consistent. I think it's best in the python/gem5/boards/*.py file, but that's just me.

File: configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py (hunk 4 above)
Comment (powerjg) [reply configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py]: There's exits for both *instructions* and for *ticks*. Which one is the goal here?

There should be just one hypercall/exit number for all tick-based exits (or we should be consistent). So, whatever we use in the simulator should be consistent.

We never want to use the `m5` module in these example standard library scripts, if at all possible. If we need to add a new API to the simulator, we should do that.

Comment (erin-le) [reply configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py]: > There's exits for both instructions and for ticks. Which one is the goal here?

Hey Jason, sorry for the mixup; I should have been looking at `Simulator.set_max_ticks()` instead of `Simulator.schedule_max_insts()`. 

> We never want to use the m5 module in these example standard library scripts, if at all possible. If we need to add a new API to the simulator, we should do that.

We will probably need to add another API to Simulator. Currently, it doesn't look like there's a function that allows access to `m5.scheduleTickExitFromCurrent` or `m5.scheduleTickExitAbsolute`.

File: configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py (hunk 3 above)
Comment (erin-le) [reply configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py]: Issue with all TODOs from this PR here: https://github.com/gem5/gem5/issues/2453

File: configs/example/gem5_library/riscv-fs.py (hunk 11 above)
Comment (erin-le) [reply configs/example/gem5_library/riscv-fs.py]: Link to PR to deprecate `riscv-fs.py`: https://github.com/gem5/gem5/pull/2452

File: configs/example/gem5_library/checkpoints/simpoints-se-restore.py (hunk 6 above)
Comment (erin-le) [reply configs/example/gem5_library/checkpoints/simpoints-se-restore.py]: I misread `insts` as `ticks` in the original function and tried switching it to m5.scheduleTickExitAbsolute as a result. We will probably have to modify the Simulator API if we want to use hypercalls for this, as `schedule_max_insts` results in a MAX_INSTS classic exit event.

Comment (powerjg) [reply configs/example/gem5_library/checkpoints/simpoints-se-restore.py]: I think it would be OK to update the simulator to use the "new" event as long as we make a note of it in the release notes

File: configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py (hunk 4 above)
Comment (powerjg) [reply configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py]: Yes, we should add that API if it doesn't exist.

Comment (erin-le) [reply configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py]: PR adding API here: https://github.com/gem5/gem5/pull/2459

File: configs/example/gem5_library/checkpoints/simpoints-se-restore.py (hunk 6 above)
Comment (erin-le) [reply configs/example/gem5_library/checkpoints/simpoints-se-restore.py]: I opened a draft PR trying to update `schedule_max_insts()` to use hypercalls here: https://github.com/gem5/gem5/pull/2514, but it doesn't work as intended.

**Comment by erin-le** [comment]
At this point, all of the configs except for `x86-mi300x-gpu.py`, `x86-spec-cpu2006-benchmarks.py`, and `x86-spec-cpu2017-benchmarks.py` have been tested. This PR is ready for review. Additional concerns I came across while testing have been recorded in a second comment in issue https://github.com/gem5/gem5/issues/2453.

**Comment by powerjg** [comment]
It would be great if @Harshil2107 could review

**Review by BobbyRBruce — APPROVED** [review]

**Review by BobbyRBruce — APPROVED** [review]
